BridgeSDK/BridgeAPI/SBBActivityManager.m

```diff
@@ -65,75 +67,99 @@ - (NSURLSessionDataTask *)getScheduledActivitiesForDaysAhead:(NSInteger)daysAhea
+    NSDate *todayEnd = [cal startOfDayForDate:[NSDate dateWithTimeIntervalSinceNow:kSBB24Hours]];
+    NSDate *windowEnd = [cal startOfDayForDate:[todayEnd dateByAddingTimeInterval:daysAhead * kSBB24Hours]];
+    
+    // things that either expired during the daysBefore period, or expire after that but start before the end of daysAhead
+    NSPredicate *predicate = [NSPredicate predicateWithFormat:@"(%K > %@ AND %K < %@) OR (%K >= %@ AND %K < %@)",
+                              NSStringFromSelector(@selector(expiresOn)), windowStart,
+                              NSStringFromSelector(@selector(expiresOn)), todayStart,
+                              NSStringFromSelector(@selector(expiresOn)), todayStart,
+                              NSStringFromSelector(@selector(scheduledOn)), windowEnd];
+    NSArray *filtered = [tasks filteredArrayUsingPredicate:predicate];
+    if (excludeValid) {
+        // "valid things" are defined as the subset of the above things that the server would return.
+        // valid things expire after the start of today and are not marked as finished. the flag is
+        // set so we want to exclude those things.
+        predicate = [NSPredicate predicateWithFormat:@"NOT (%K > %@ AND %K == nil)",
+                     NSStringFromSelector(@selector(expiresOn)), todayStart,
+                     NSStringFromSelector(@selector(finishedOn))
+                     ];
+        filtered = [filtered filteredArrayUsingPredicate:predicate];
+    }
+    
+    return filtered;
 }
 
 - (NSURLSessionDataTask *)getScheduledActivitiesForDaysAhead:(NSInteger)daysAhead daysBehind:(NSInteger)daysBehind cachingPolicy:(SBBCachingPolicy)policy withCompletion:(SBBActivityManagerGetCompletionBlock)completion
 {
     NSMutableDictionary *headers = [NSMutableDictionary dictionary];
     [self.authManager addAuthHeaderToHeaders:headers];
     
-    __block SBBResourceList *tasks = nil;
-    NSArray *daysBehindTasks = [NSArray array];
+    NSArray *savedTasks = [NSArray array];
     id<SBBCacheManagerProtocol> cacheManager = nil;
-    if (policy != SBBCachingPolicyNoCaching) {
-        if ([self.objectManager conformsToProtocol:@protocol(SBBObjectManagerInternalProtocol)]) {
+    if (gSBBUseCache) {
+        BOOL isInternalObjectManager = [self.objectManager conformsToProtocol:@protocol(SBBObjectManagerInternalProtocol)];
+        if (isInternalObjectManager) {
             cacheManager = ((id<SBBObjectManagerInternalProtocol>)self.objectManager).cacheManager;
         }
-        tasks = (SBBResourceList *)[cacheManager cachedSingletonObjectOfType:@"ResourceList" createIfMissing:NO];
+        SBBResourceList *tasks = (SBBResourceList *)[cacheManager cachedSingletonObjectOfType:@"ResourceList" createIfMissing:NO];
         
         // if we're going straight to cache, we're done
         if (policy == SBBCachingPolicyCachedOnly) {
             dispatch_async(dispatch_get_main_queue(), ^{
```

Comment:
I'm... not entirely sure what I was thinking here. If I wanted to just get it off onto another queue so the caller didn't block while the handler ran, then the main queue is probably exactly the *wrong* queue to do it in. The "normal" call to this method's handler would be called from the network request completion handler, which is documented to run on the "delegate queue" for the NSURLSession but since it's the default session and you can't set a delegate on it, I'm not entirely sure what that would mean, or if it would even *have* a delegate queue. *Shrug.* I'll take it out; it doesn't really seem necessary.